### PR TITLE
New Commit

### PR DESCRIPTION
**Pull request by Tolyaspit**
Please check changes and approve the request!

### CODE REVIEW

File: src/Abstract.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+public abstract class Abstract {
+    private int id;
+
+//    public int getId(){
+//        return id;
+//    }
```
Comment (Alexmansar) [inline src/Abstract.java:6]: uncommented it please

File: src/Abstract.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
+public abstract class Abstract {
```
Comment (Alexmansar) [inline src/Abstract.java:1]: AbstractId for example or AbstractNumericId

File: src/Admin.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    public void printRole() {
+        System.out.println("Role: Admin");
+    }
+
+    @Override
+    public void print() {
+
+    }
+
+    public void checkTicket(Ticket ticket) {
```
Comment (Alexmansar) [inline src/Admin.java:12]: should return some boolean ( according to the name)

File: src/Abstract.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+    private int id;
```
Comment (Alexmansar) [inline src/Abstract.java:2]: make it protected)

File: src/Ticket.java [hunk 5]
```diff
@@ -1,8 +1,9 @@
-public class Ticket {
+public class Ticket extends Abstract{
     private String id;
```
Comment (Alexmansar) [inline src/Ticket.java:2]: we don't need it already (see parent class)

File: src/Ticket.java [hunk 6]
```diff
@@ -31,11 +32,11 @@ public Ticket(String concertHall, String eventCode, UnixTimePeriod timePeriod) {
...
-    public String getId(){
+    public String getId() {
         return id;
     }
 
-    public void setId(String id){
+    private void setId(String id){
         if ((id.length()==4) && id.matches(("[a-zA-Z0-9]{4}"))) {
             this.id = id;
         }else {
```
Comment (Alexmansar) [inline src/Ticket.java:42]: move to the Abstract class

File: src/TicketService.java [hunk 7]
```diff
@@ -14,6 +14,17 @@ public static void main(String[] args) {
 
         List<Ticket> ticketsFromStadiumSector = ticketService.getTicketsByStadiumSector(Ticket.StadiumSector.A);
         ticketService.printTickets(ticketsFromStadiumSector);
+
+        Client client = new Client();
+        client.printRole();
+        Ticket clientTicket = client.getTicket("A123",ticketService);
+
+        Admin admin = new Admin();
```
Comment (Alexmansar) [inline src/TicketService.java:22]: User admin = new Admin()

File: src/Admin.java [hunk 8]
```diff
@@ -0,0 +1,15 @@
+public class Admin extends User{
+    @Override
+    public void printRole() {
+        System.out.println("Role: Admin");
+    }
+
+    @Override
+    public void print() {
+
```
Comment (Alexmansar) [inline src/Admin.java:9]: add some logic

**Review by Alexmansar — CHANGES_REQUESTED** [review]
@Tolyaspit Hi, please check some comments

File: src/Ticket.java [hunk 9]
```diff
@@ -32,16 +32,14 @@ public Ticket(String concertHall, String eventCode, UnixTimePeriod timePeriod) {
...
-    private void setId(String id){
-        if ((id.length()==4) && id.matches(("[a-zA-Z0-9]{4}"))) {
-            this.id = id;
-        }else {
-            throw new IllegalArgumentException("ID must be exactly 4 digits or characters");
-        }
+    @Override
+    public void setId(String id) {
+        super.setId(id);
     }
```
Comment (Alexmansar) [inline src/Ticket.java:43]: you don't need it. You will use just parent methods

File: src/Ticket.java [hunk 10]
```diff
@@ -95,18 +98,28 @@ public double getMaxAllowedBackpackWeight() {
...
 
+    @Override
+    public boolean equals(Object obj) {
+        return super.equals(obj);
+    }
+
+    @Override
+    public int hashCode() {
+        return super.hashCode();
+    }
```
Comment (Alexmansar) [inline src/Ticket.java:121]: which super you use here?

**Review by Alexmansar — CHANGES_REQUESTED** [review]
@Tolyaspit please, check small comments))

File: src/Ticket.java (hunk 10 above)
Comment (Tolyaspit) [reply src/Ticket.java:121]: Oh , I used the default shortcut. Do I need to add the fields to compare for these override methods?

File: src/Ticket.java [hunk 11]
```diff
@@ -95,18 +90,31 @@ public double getMaxAllowedBackpackWeight() {
...
+    private void setPrice(double ticketPrice) {
         this.ticketPrice = ticketPrice;
     }
 
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Ticket ticket = (Ticket) o;
+        return time == ticket.time && isPromo == ticket.isPromo && Double.compare(maxAllowedBackpackWeight, ticket.maxAllowedBackpackWeight) == 0 && Double.compare(ticketPrice, ticket.ticketPrice) == 0 && Objects.equals(concertHall, ticket.concertHall) && Objects.equals(eventCode, ticket.eventCode) && timePeriod == ticket.timePeriod && stadiumSector == ticket.stadiumSector;
```
Comment (Alexmansar) [inline src/Ticket.java:110]: please, start some code from new line))
```
return time == ticket.time 
   && isPromo == ticket.isPromo 
   && Double.compare(maxAllowedBackpackWeight, ticket.maxAllowedBackpackWeight) == 0 
   && Double.compare(ticketPrice, ticket.ticketPrice) == 0 
   && Objects.equals(concertHall, ticket.concertHall)
   && Objects.equals(eventCode, ticket.eventCode) 
   && timePeriod == ticket.timePeriod 
   && stadiumSector == ticket.stadiumSector;
```
good practice)

**Review by Alexmansar — APPROVED** [review]
@Tolyaspit Could be merged. Please just fix styling and merge))

File: src/Ticket.java (hunk 11 above)
Comment (Tolyaspit) [reply src/Ticket.java:110]: Ok:)thanks

**Pull request merged**